crates/motoko/src/lib/shared.rs

```diff
@@ -0,0 +1,51 @@
+use core::ops::Deref;
+use std::rc::Rc;
+
+use serde::{Deserialize, Serialize};
+
+#[derive(Debug, Clone, PartialEq, Eq, Hash, Serialize, Deserialize)]
+pub struct Shared<T> {
+    rc: Rc<T>,
+}
+
+impl<T: Clone> Shared<T> {
+    /// Call `value.share()` to construct a `Shared` value.
+    fn new(x: T) -> Shared<T> {
+        Shared { rc: Rc::new(x) }
+    }
+
+    /// A more explicit alternative to `clone()` (same use case as calling `Rc::clone(_)`).
+    pub fn fast_clone(&self) -> Self {
+        self.clone()
+    }
+
+    #[inline(always)]
+    pub fn get(&self) -> T {
+        self.rc.deref().clone()
+    }
+}
+
+impl<T> Deref for Shared<T> {
+    type Target = T;
+
+    #[inline(always)]
+    fn deref(&self) -> &T {
+        self.rc.deref()
+    }
+}
+
+impl<T> AsRef<T> for Shared<T> {
+    fn as_ref(&self) -> &T {
+        self.rc.as_ref()
+    }
+}
+
+pub trait Share<T> {
+    fn share(self) -> Shared<T>;
```

Comment:
Oh, I see that we are also using the name `share`.

Then why `fast_clone` also?

Reply:
`share()` turns `T` into `Shared<T>` (similar to `Rc::new(_)`), while `fast_clone()` is just an alias for `clone()` so we can visually distinguish these from "slow clones" which otherwise use `.clone()`.

Reply:
Ah, I think I see now.  I was projecting my own internal meanings, and not looking close enough.

`share()` is actually to be avoided if possible, since it indicates a monadic operation in the `Shared<_>` space -- constructing new data that actually is not yet shared.

By contrast, `fast_clone()` is to be embraced, since it's like `Rc::clone(_)` -- an actual point of "new sharing",  in O(1) time.

Does that match your thinking?